File: revised/Combination Sum II.cpp
/**
 * dfs + 标记数组避免重复
 */
class Solution {
    vector<vector<int> > ret_;
    vector<int> node_;
public:
    void dfs(vector<int> &candidates,int target,int cur,int level){
        unordered_set<int> tag ; 
        for(int i = level ; i< candidates.size() ; i++ ) {
            if(tag.count(candidates[i]) >0){
                continue;
            }
            tag.insert(candidates[i]);
            if(cur + candidates[i] > target ){
                break;
            }else if(cur + candidates[i] < target){
                node_.push_back(candidates[i]);
                dfs(candidates , target, cur+candidates[i],i+1) ;
                node_.pop_back();
            }else{
                node_.push_back(candidates[i]);
                ret_.push_back(node_);
                node_.pop_back();
            }

        }
    }
    vector<vector<int> > combinationSum2(vector<int> &candidates, int target) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        sort(candidates.begin(),candidates.end());
        ret_.clear();
        node_.clear();
        dfs(candidates,target,0,0); 
        return ret_;
    }
};

File: revised/Combination Sum II.java
/**
Given a collection of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.

Each number in C may only be used once in the combination. 
注意
All numbers (including target) will be positive integers.
Elements in a combination (a1, a2, … , ak) must be in non-descending order. (ie, a1 ≤ a2 ≤ … ≤ ak).
The solution set must not contain duplicate combinations.
*/
public class Solution {
    /**
     * @param num: Given the candidate numbers
     * @param target: Given the target number
     * @return: All the combinations that sum to target
     */
     List<List<Integer>> cs = null; 
     int []flag  = null; 
    public List<List<Integer>> combinationSum2(int[] num, int target) {
        // write your code here
        cs = new ArrayList<List<Integer>> () ;
        if(num.length == 0 ) return cs; 
        Arrays.sort(num) ; 
        flag = new int[num.length]; Arrays.fill(flag, 0) ; 
        dfs(num,target, 0 ) ;  
        return cs; 
    }
    public void dfs(int []num , int target , int s ){
        if(target < 0 ) return ; 
        if(target == 0 ){
            List<Integer> row = new ArrayList<Integer>();
            for(int i = 0 ;i <num.length ; i++ ){
                if(flag[i] == 1 ) {
                    row.add(num[i]) ; 
                }
            }
            cs.add(row); 
            return ; 
        }
        for(int i = s ; i< num.length ; i++){
            if(i>s && num[i] == num[i-1] ) { // avoid duplicated combinations
                continue; 
            }
            flag[i] = 1 ; 
            dfs(num , target - num[i] , i + 1 ) ; 
            flag[i] = 0 ; 
        }
    }
    
}

